crates/ar-cli/src/collect_viewing_keys.rs

```diff
@@ -0,0 +1,235 @@
+use std::{
+    fs::File,
+    io::{self, Read},
+};
+
+use alloy_json_rpc::RpcError;
+use alloy_network::AnyNetwork;
+use alloy_primitives::{Address, U256};
+use alloy_provider::{Provider, ProviderBuilder};
+use alloy_rpc_types::{BlockNumberOrTag, BlockTransactionsKind, TransactionTrait};
+use alloy_sol_types::SolCall;
+use alloy_transport::TransportErrorKind;
+use hex::FromHexError;
+use log::{debug, info};
+use rusqlite::Connection;
+use shielder_circuits::{grumpkin, GrumpkinPointAffine};
+use shielder_contract::{
+    ShielderContract::{newAccountERC20Call, newAccountNativeCall},
+    ShielderContractError,
+};
+use thiserror::Error;
+use type_conversions::u256_to_field;
+
+use crate::{
+    cli::Endianess,
+    db::{self, ViewingKey},
+};
+
+#[derive(Debug, Error)]
+#[error(transparent)]
+#[non_exhaustive]
+pub enum CollectKeysError {
+    #[error("Error while interacting with the Shielder contract")]
+    Contract(#[from] ShielderContractError),
+
+    #[error("RPC error")]
+    Rpc(#[from] RpcError<TransportErrorKind>),
+
+    #[error("Hex decoding error")]
+    HexError(#[from] FromHexError),
+
+    #[error("Error reading AR private key file")]
+    ARKeyRead(#[from] std::io::Error),
+
+    #[error("Error converting from a little-endian byte representation to grumpkin::Fr")]
+    NotAGrumpkinBaseFieldElement,
+
+    #[error("Error while persisting data")]
+    Db(#[from] rusqlite::Error),
+}
+
+/// Goes back in transaction history and collects all viewing keys
+/// - look for new_account txs
+/// - read c1,c2
+/// - decrypt message  => k (viewing key)
+/// - record k
+pub async fn run(
+    rpc_url: &str,
+    shielder_address: &Address,
+    private_key_file: &str,
+    endianess: Endianess,
+    from_block: u64,
+    connection: Connection,
+) -> Result<(), CollectKeysError> {
+    let provider = ProviderBuilder::new()
+        .network::<AnyNetwork>()
+        .on_builtin(rpc_url)
+        .await?;
+
+    let last_finalized_block_number = provider.get_block_number().await?;
+
+    let bytes = match endianess {
+        Endianess::LitteEndian => private_key_bytes(private_key_file)?,
+        Endianess::BigEndian => {
+            let mut bytes = private_key_bytes(private_key_file)?;
+            bytes.reverse();
+            bytes
+        }
+    };
+
+    let private_key = grumpkin::Fr::from_bytes(&bytes)
+        .into_option()
+        .ok_or(CollectKeysError::NotAGrumpkinBaseFieldElement)?;
+
+    db::create_viewing_keys_table(&connection)?;
+    // TODO: checkpoint tx blocks
+
+    for block_number in from_block..=last_finalized_block_number {
+        if let Some(block) = provider
+            .get_block_by_number(
+                BlockNumberOrTag::Number(block_number),
+                BlockTransactionsKind::Full,
+            )
+            .await?
+        {
+            if let Some(txs) = block.transactions.as_transactions() {
+                for tx in txs {
```

Comment:
here the same suggestions - we can log lack of transactions; also, we can use `block.transactions.as_transactions().unwrap_or_else(|| log!(no txs); vec![])`